Regression with GitLab for updated openid-client v6

Hi

We are using the `ep_openid_connect` with GitLab as the OIDC provider. In the recent update from `v3.0.21` to `v.3.0.22`, the authentication stopped working with the following error:

```
$ pnpm --filter ep_etherpad-lite run prod
$ cross-env NODE_ENV=production node --require tsx/cjs node/server.ts
[...]
[2026-05-11T16:24:52.065] [INFO] plugins - Loading plugin ep_openid_connect...
[...]
[2026-05-11T16:25:10.069] [INFO] server - Installed plugins: ep_plugin_helpers@0.2.9, ep_headings2@0.2.117, ep_delete_after_delay@0.0.36, ep_openid_connect@3.0.22, ep_user_displayname@1.0.44
[...]
[2026-05-11T16:25:10.381] [INFO] plugin:ep_openid_connect - OpenID Connect Discovery complete.
[2026-05-11T16:25:10.381] [INFO] plugin:ep_openid_connect - Configured.
[...]
[2026-05-12T06:23:53.749] [INFO] http - Failed authentication from IP 172.16.0.54
[2026-05-12T06:23:54.478] [ERROR] settings - WWWAuthenticateChallengeError: server responded with a challenge in the WWW-Authenticate HTTP Header
    at checkAuthenticationChallenges (/opt/etherpad-lite/src/plugin_packages/.versions/oauth4webapi@3.8.6/src/index.ts:3685:11)
    at checkOAuthBodyError (/opt/etherpad-lite/src/plugin_packages/.versions/oauth4webapi@3.8.6/src/index.ts:2830:5)
    at processGenericAccessTokenResponse (/opt/etherpad-lite/src/plugin_packages/.versions/oauth4webapi@3.8.6/src/index.ts:3571:9)
    at processAuthorizationCodeOpenIDResponse (/opt/etherpad-lite/src/plugin_packages/.versions/oauth4webapi@3.8.6/src/index.ts:4085:24)
    at Module.processAuthorizationCodeResponse (/opt/etherpad-lite/src/plugin_packages/.versions/oauth4webapi@3.8.6/src/index.ts:4031:12)
    at Module.authorizationCodeGrant (/opt/etherpad-lite/src/plugin_packages/.versions/openid-client@6.8.4/src/index.ts:3474:19)
    at process.processTicksAndRejections (node:internal/process/task_queues:103:5)
    at async /opt/etherpad-lite/src/plugin_packages/.versions/ep_openid_connect@3.0.22/index.js:174:22
```

This is directly related to the manual migration changes in https://github.com/ether/ep_openid_connect/pull/139. GitLab advertises both `client_secret_basic` and `client_secret_post` in its [`.well-known/openid-configuration`](https://gitlab.com/.well-known/openid-configuration), but apparently the `client_secret_basic` has an incompatibility right now.

I suggest making this field configurable, like many of the other behaviors in the OpenID Connect flow already are via the Etherpad configuration file. Then we can manually switch to the `client_secret_post` behavior again, which mimics the behavior present in `ep_openid_connect` > `v3.0.22`.

--- a/README.md
+++ b/README.md
@@ -49,6 +49,11 @@ Etherpad's `requireAuthentication` setting must be `true`.
 * `client_id` (required): The OAuth2 client ID issued by the identity provider.
 * `client_secret` (required): The OAuth2 client secret issued by the identity
   provider.
+* `token_endpoint_auth_method` (optional; defaults to `client_secret_basic`):
+  Client authentication method to use for token endpoint requests. Supported
+  values are `client_secret_basic` and `client_secret_post`. Use
+  `client_secret_post` if your identity provider expects the client secret in
+  the token request body instead of the HTTP `Authorization` header.
 * `base_url` (required): The public base Etherpad URL. When registering Etherpad
   with your identity provider, the redirect URL (a.k.a. callback URL) is this
   base URL plus `/ep_openid_connect/callback`.

--- a/index.js
+++ b/index.js
@@ -10,6 +10,7 @@ for (const level of ['debug', 'info', 'warn', 'error']) {
 const defaultSettings = {
   prohibited_usernames: ['admin', 'guest'],
   scope: ['openid'],
+  token_endpoint_auth_method: 'client_secret_basic',
   user_properties: {},
 };
 let settings;
@@ -33,6 +34,7 @@ const validSettings = new Ajv().compile({
     issuer_metadata: {},
     prohibited_usernames: {elements: {type: 'string'}},
     scope: {elements: {type: 'string'}},
+    token_endpoint_auth_method: {enum: ['client_secret_basic', 'client_secret_post']},
     user_properties: {values: {
       optionalProperties: {
         claim: {type: 'string'},
@@ -100,11 +102,25 @@ const discoverConfig = async (issuerUrl, clientId, clientAuth) => {
   }
 };
 
+const getClientAuth = (settings) => {
+  switch (settings.token_endpoint_auth_method) {
+    case 'client_secret_basic':
+      // eslint-disable-next-line new-cap
+      return oidc.ClientSecretBasic(settings.client_secret);
+    case 'client_secret_post':
+      // eslint-disable-next-line new-cap
+      return oidc.ClientSecretPost(settings.client_secret);
+    default:
+      throw new Error(
+          `Unsupported token endpoint auth method: ${settings.token_endpoint_auth_method}`);
+  }
+};
+
 const buildConfig = async (settings) => {
   // openid-client@5 defaulted the token endpoint auth method to
   // `client_secret_basic`; v6 defaults to `client_secret_post`. Pin the
-  // method explicitly so existing IdP registrations keep working.
-  const clientAuth = oidc.ClientSecretBasic(settings.client_secret);
+  // default method explicitly so existing IdP registrations keep working.
+  const clientAuth = getClientAuth(settings);
   if (settings.issuer) {
     const config =
         await discoverConfig(settings.issuer, settings.client_id, clientAuth);

--- a/static/tests/backend/specs/tests.js
+++ b/static/tests/backend/specs/tests.js
@@ -125,6 +125,31 @@ describe(__filename, function () {
     assert.equal(msg.data.readonly, false);
   });
 
+  describe('token_endpoint_auth_method', function () {
+    for (const tokenEndpointAuthMethod of ['client_secret_basic', 'client_secret_post']) {
+      it(`supports ${tokenEndpointAuthMethod}`, async function () {
+        await epOpenidConnect.loadSettings('loadSettings', {settings: {ep_openid_connect: {
+          ...pluginSettings,
+          token_endpoint_auth_method: tokenEndpointAuthMethod,
+        }}});
+        const padId = common.randomString();
+        const url = new URL(`/p/${padId}`, common.baseUrl).toString();
+        const res = await login(agent, issuer, url, 'normalUser');
+        assert.equal(res.request.url, url);
+        assert.equal(res.status, 200);
+      });
+    }
+
+    it('rejects unsupported methods', async function () {
+      await epOpenidConnect.loadSettings('loadSettings', {settings: {ep_openid_connect: {
+        ...pluginSettings,
+        token_endpoint_auth_method: 'unsupported',
+      }}});
+      await agent.get(new URL('/ep_openid_connect/login', common.baseUrl))
+          .expect(401);
+    });
+  });
+
   it('noCreate user is unable to create a pad', async function () {
     const padId = common.randomString();
     const url = new URL(`/p/${padId}`, common.baseUrl).toString();

